pkg/rke2/psp/psp.go

```diff
@@ -0,0 +1,443 @@
+//
+// Load logic
+//
+// CIS:
+// - If the globalRestricted annotation does not exist, create PSP, role, binding.
+// - If the systemUnrestricted annotation does not exist, create the PSP, role, binding.
+// - If the globalUnrestricted annotation does not exist, check if PSP exists, and if so,
+//   delete it, the role, and the bindings.
+//
+// non-CIS:
+// - If the globalUnrestricted annotation does not exist, then create PSP, role, binding.
+// - If the systemUnrestricted annotation does not exist, then create PSP, role, binding.
+// - If the globalRestricted annotation does not exist, then check if the PSP exists and
+//   if it doesn't, create it. Check if the associated role and bindings exist and
+//   if they do, delete them.
+func SetPSPs(sCtx context.Context, ctx *cli.Context, k8sWrapTransport transport.WrapperFunc) error {
+	const (
+		kubeConfigPath = "/etc/rancher/rke2/rke2.yaml"
+		waitDelay      = time.Second * 1
+	)
+
+	// wait until the kubeconfig file is written
+	for !kubeConfigExists(kubeConfigPath) {
+		time.Sleep(waitDelay)
+	}
+
+	cs, err := newClient(kubeConfigPath, k8sWrapTransport)
+	if err != nil {
+		return err
+	}
+
+	for {
+		// wait until kube-apiserver is running
+		if _, err := cs.Discovery().ServerVersion(); err != nil {
+			time.Sleep(waitDelay)
+			continue
+		}
+
+		// wait until the kube-system namespace is created
+		ns, err := cs.CoreV1().Namespaces().Get(sCtx, kubeSystemNamespace, metav1.GetOptions{})
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				time.Sleep(waitDelay)
+				continue
+			}
+			return err
+		}
+
+		if ctx.String("profile") == "" { // non-CIS mode
+			if err := setGlobalUnrestricted(sCtx, cs, ns); err != nil {
+				return err
+			}
+
+			if err := setSystemUnrestricted(sCtx, cs, ns); err != nil {
+				return err
+			}
+
+			if _, ok := ns.Annotations[namespaceAnnotationGlobalRestricted]; !ok {
+				// check if the policy exists
+				if _, err := cs.PolicyV1beta1().PodSecurityPolicies().Get(sCtx, globalRestrictedPSPName, metav1.GetOptions{}); err != nil {
+					if apierrors.IsNotFound(err) {
+						tmpl := fmt.Sprintf(globalRestrictedPSPTemplate, globalRestrictedPSPName)
+						if err := deployPodSecurityPolicyFromYaml(sCtx, cs, tmpl); err != nil {
+							return err
+						}
+					}
+				}
+
+				// check if role exists and delete it
+				role, err := cs.RbacV1().ClusterRoles().Get(sCtx, globalRestrictedRoleName, metav1.GetOptions{})
+				if err != nil {
+					if apierrors.IsNotFound(err) {
+						logrus.Info(err)
+					} else {
+						if err := cs.RbacV1().ClusterRoles().Delete(sCtx, role.Name, metav1.DeleteOptions{}); err != nil {
+							return err
+						}
+					}
+				}
+
+				// check if role binding exists and delete it
+				roleBinding, err := cs.RbacV1().ClusterRoleBindings().Get(sCtx, globalRestrictedRoleBindingName, metav1.GetOptions{})
+				if err != nil {
+					if apierrors.IsNotFound(err) {
+						logrus.Info(err)
```

Comment:
I wouldn't even log this error. This is "normal" state after the initial boostrapping, so isNotFound can just be swallowed

Reply:
Refactored block:

```go
role, err := cs.RbacV1().ClusterRoles().Get(sCtx, globalRestrictedRoleName, metav1.GetOptions{})
				if err != nil && !apierrors.IsNotFound(err) {
					return err
				}
				if err := cs.RbacV1().ClusterRoles().Delete(sCtx, role.Name, metav1.DeleteOptions{}); err != nil {
					return err
				}
```

Reply:
looks good, though i make another comment further down that to avoid race conditions you should check if the error post delete is due to the resource not existing and if thats the case, ignore the error